File: packages/flutter-client/lib/src/auth/auth_lens.dart
import 'package:collection/collection.dart';

/// Defines the authentication context attached to Firestore operations.
sealed class AuthLens {
  const AuthLens();

  Map<String, dynamic> toMap();

  String get mode {
    switch (this) {
      case AdminLens():
        return 'admin';
      case AnonLens():
        return 'anon';
      case AppSessionLens():
        return 'app-session';
      case AsUserLens():
        return 'as';
    }
  }

  String? get uid => this is AsUserLens ? (this as AsUserLens).uid : null;
  String? get tenant => this is AsUserLens ? (this as AsUserLens).tenant : null;
  Map<String, dynamic>? get token =>
      this is AsUserLens ? (this as AsUserLens).token : null;

  static const AuthLens anon = AnonLens();
  static const AuthLens admin = AdminLens();
  static const AuthLens appSession = AppSessionLens();

  static AuthLens asUser({
    required String uid,
    String? tenant,
    Map<String, dynamic>? token,
  }) =>
      AsUserLens(uid: uid, tenant: tenant, token: token);
}

/// Unauthenticated lens — forces Security Rules to evaluate with `request.auth == null`.
class AnonLens extends AuthLens {
  const AnonLens();

  @override
  Map<String, dynamic> toMap() => const {'mode': 'anon'};

  @override
  bool operator ==(Object other) => other is AnonLens;

  @override
  int get hashCode => 0;

  @override
  String toString() => 'AuthLens.anon';
}

/// Admin bypass lens — bypasses Security Rules checks entirely.
class AdminLens extends AuthLens {
  const AdminLens();

  @override
  Map<String, dynamic> toMap() => const {'mode': 'admin'};

  @override
  bool operator ==(Object other) => other is AdminLens;

  @override
  int get hashCode => 1;

  @override
  String toString() => 'AuthLens.admin';
}

/// Browser session lens — inherits the browser tab's interactive identity.
class AppSessionLens extends AuthLens {
  const AppSessionLens();

  @override
  Map<String, dynamic> toMap() => const {'mode': 'app-session'};

  @override
  bool operator ==(Object other) => other is AppSessionLens;

  @override
  int get hashCode => 2;

  @override
  String toString() => 'AuthLens.appSession';
}

/// Authenticated user impersonation lens — evaluates Security Rules as the given identity.
class AsUserLens extends AuthLens {
  @override
  final String uid;
  @override
  final String? tenant;
  @override
  final Map<String, dynamic>? token;

  const AsUserLens({required this.uid, this.tenant, this.token});

  @override
  Map<String, dynamic> toMap() => {
        'mode': 'as',
        'uid': uid,
        if (tenant != null) 'tenant': tenant,
        if (token != null) 'token': token,
      };

  @override
  bool operator ==(Object other) =>
      other is AsUserLens &&
      other.uid == uid &&
      other.tenant == tenant &&
      const DeepCollectionEquality().equals(other.token, token);

  @override
  int get hashCode => Object.hash(
        uid,
        tenant,
        const DeepCollectionEquality().hash(token),
      );

  @override
  String toString() => 'AuthLens.asUser(uid: $uid, tenant: $tenant, token: $token)';
}

File: packages/kt-client/src/main/kotlin/dev/pyric/auth/AuthLens.kt
package dev.pyric.auth

sealed class AuthLens {
    abstract fun toMap(): Map<String, Any?>

    data object Anon : AuthLens() {
        override fun toMap(): Map<String, Any?> = mapOf("mode" to "anon")
    }

    data object Admin : AuthLens() {
        override fun toMap(): Map<String, Any?> = mapOf("mode" to "admin")
    }

    data object AppSession : AuthLens() {
        override fun toMap(): Map<String, Any?> = mapOf("mode" to "app-session")
    }

    data class AsUser(
        val uid: String,
        val token: Map<String, Any?>? = null,
        val tenant: String? = null
    ) : AuthLens() {
        override fun toMap(): Map<String, Any?> {
            val tokenMap = if (token != null) {
                val merged = mutableMapOf<String, Any?>("sub" to uid, "user_id" to uid)
                merged.putAll(token)
                merged
            } else {
                mapOf("sub" to uid, "user_id" to uid)
            }
            val map = mutableMapOf<String, Any?>(
                "mode" to "as",
                "uid" to uid,
                "token" to tokenMap
            )
            if (tenant != null) map["tenant"] = tenant
            return map
        }
    }

    companion object {
        fun fromMap(map: Map<String, Any?>): AuthLens {
            val mode = (map["mode"] as? String)?.lowercase() ?: "anon"
            return when (mode) {
                "admin" -> Admin
                "app-session", "app_session" -> AppSession
                "as", "user" -> {
                    val uid = map["uid"] as? String ?: ""
                    @Suppress("UNCHECKED_CAST")
                    val token = map["token"] as? Map<String, Any?>
                    val tenant = map["tenant"] as? String
                    AsUser(uid = uid, token = token, tenant = tenant)
                }
                "anon", "anonymous" -> Anon
                else -> Anon
            }
        }
    }
}
